File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        prevNode = None
        temp = head
        
        while temp:
            kthNode = self.getKthNode(temp, k)
            if not kthNode:
                prevNode.next = temp
                break
            nxtNode = kthNode.next
            kthNode.next = None
            self.reverse(temp)
            
            if temp == head:
                head = kthNode
            else:
                prevNode.next = kthNode
            prevNode = temp
            temp = nxtNode
        return head
    
    def reverse(self, node):
        prev = temp = None
        
        while node:
            temp = node.next
            node.next = prev
            prev = node
            node = temp
    
    def getKthNode(self, node, k):
        k -= 1
        
        while node and k > 0:
            k -= 1
            node = node.next
        
        return node
        

File: 0025-reverse-nodes-in-k-group/0025-reverse-nodes-in-k-group.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode temp = head;
        ListNode prevNode = null;
        
        while (temp != null){
            ListNode kthNode = getkthNode(temp, k);
            if (kthNode == null){
                prevNode.next = temp;
                break;
            }
    
            ListNode nxtNode = kthNode.next;
            kthNode.next = null;
            reverse(temp);
            if (temp == head) head = kthNode;
            else{
                prevNode.next = kthNode;    
            }
            prevNode = temp;
            temp = nxtNode;
        }
        return head;
        
    }
    
    private void reverse(ListNode node){
        ListNode prev = null, temp = null;
        
        while (node != null){
            temp = node.next;
            node.next = prev;
            prev = node;
            node = temp;
        }
    }
    private ListNode getkthNode(ListNode node, int k){
        k--;
        
        while ((node != null) && (k > 0)){
            k--;
            node = node.next;
        }
        return node;
    }
}